### ISSUE TITLE
StringComparer.InvariantCulture.GetHashCode throws in Windows 7 compat mode

### ISSUE THREAD

**Issue by vitek-karas** [dotnet/coreclr#26479]
#### Repro
`dotnet new console`
Use this code in `Main`
```C#
            string value = @"Some";
            IEqualityComparer<string> comparer = StringComparer.InvariantCulture;
            Console.WriteLine($"{comparer.GetHashCode(value)}");
```
Build with .NET Core 3.0.
Go to the output folder and set compatibility mode for the `.exe` to `Windows 7`.
Run the exe.

#### Output
```
Unhandled exception. System.ArgumentException: External component has thrown an exception.
   at System.Globalization.CompareInfo.GetHashCodeOfStringCore(ReadOnlySpan`1 source, CompareOptions options)
   at System.Globalization.CompareInfo.GetHashCodeOfString(String source, CompareOptions options)
   at System.CultureAwareComparer.GetHashCode(String obj)
   at ConsoleApp3.Program.Main(String[] args)
```

This seems to happen for any string other than empty string. The program works fine when executed without the compatibility mode set.

I haven't tried to run this on actual Windows 7 machine.

**Comment by vitek-karas** [comment]
Originally reported in https://github.com/dotnet/core-setup/issues/7872.

**Comment by jkotas** [comment]
This is duplicate of https://github.com/dotnet/coreclr/issues/15021. The Windows compatibility shims do not correctly implement `LCMAP_SORTHANDLE` flag that .NET Core depends on.

**Comment by vitek-karas** [comment]
This is basically extending the scenarios where it can happen. In .NET Core 3.0 all apps get an `.exe` by default, so it's not just setting compatibility mode on the `dotnet.exe` which I agree is something we should not support. This is about each particular app.

Also with .NET Core 3.0 this is then exaggerated by the much wider support for native hosting, which means there will be more native apps which host .NET Core for pieces of functionality. If the app itself wants to use compatibility modes, it will break. Note that this also applies to .NET Core COM servers - if a COM object which is implemented in .NET Core is ever activated by an app running in the compatibility mode, it will fail to activate.
/cc @AaronRobinsonMSFT

**Comment by Sewer56** [comment]
> Also with .NET Core 3.0 this is then exaggerated by the much wider support for native hosting, which means there will be more native apps which host .NET Core for pieces of functionality.

This is actually the original scenario that caused https://github.com/dotnet/core-setup/issues/7872 to be reported.

I specifically noticed this issue when trying to execute .NET code in a foreign native process through the use of DLL Injection of a native bootstrapper which uses the `nethost` hosting API.

The foreign native process has a small issue running on modern OS in a very specific place, which is why I had compatibility mode enabled in the first place.

![](https://raw.githubusercontent.com/Reloaded-Project/Reloaded-II/master/Docs/Diagrams/Images/Architecture-Diagram-Reloaded-II.png)

Diagram from [Reloaded-II Documentation](https://github.com/Reloaded-Project/Reloaded-II/blob/master/Docs/Reloaded-II-Architecture.md#reloaded-ii) provided for clarity.

--------------

Note: This issue isn't limited to Windows 7 compatibility mode. It will happen with Windows 8 compat mode aswell.

**Comment by tarekgh** [comment]
CC @ericstj 

This will be risky to try to fix for 3.0. This is not a new issue and existed in previous releases too. We never supported running .net core apps under the Windows comparability modes. trying to fix that will not only just fixing this case but also we'll need to do more work to ensure we run fine under the comparability modes. 

Also, the underlying issue is really a bug in Windows and not in our products. I contacted Windows team before and they decided to not fix it. I'll try to bring this issue back to them to consider. but for now, I am going to move this issue to 5.0 release and apps need to work around this by not applying the compatibility mode.

**Comment by tarekgh** [comment]
CC @danmosemsft

**Comment by danmosemsft** [comment]
Marking 3.0.x for servicing consideration. Please leave this active after merging into master.

**Comment by tarekgh** [comment]
@danmosemsft should we keep this issue active as you marked it as 3.0.x?

**Comment by danmosemsft** [comment]
Yes please let's use issues + 3.0.x milestone for now while we figure out process.

**Comment by tarekgh** [comment]
sorry I didn't see your previous comment. I''l reactivate the issue.

**Comment by tarekgh** [comment]
@danmosemsft I think this issue is not meeting 3.0.x bar but I believe having it for 3.1 is good. what you think? should I move it to 3.1?

CC @ericstj

**Comment by danmosemsft** [comment]
@tarekgh Do we run perfectly (as far as is known) in compat mode, after this fix, or are there more issues?

If this fix enables the whole scenario, and the risk is acceptable, I support porting into 3.1 - please go ahead and prep a PR.

**Comment by danmosemsft** [comment]
I ask because I see you said
>  also we'll need to do more work to ensure we run fine under the comparability modes.

In the PR could you put the usual template, even if it's just for us?

**Comment by tarekgh** [comment]
> Do we run perfectly (as far as is known) in compat mode, after this fix, or are there more issues?

I have done basic scenarios tests when I fixed the issue and it was working fine and didn't see any other issue. Before this fix, it was not possible to launch the net core app at all and now it launch fine. If there will be any issue we are not aware of, that will be localized to that area and not have global impact as the one I fixed. 

Regarding the risk, it is low as I am not changing behavior. we are just deciding if we can use the flag that can give us better perf when calling the OS.

> In the PR could you put the usual template, even if it's just for us?

Yes, I'll do that.
